pkg/stanza/operator/parser/container/parser.go

```diff
@@ -0,0 +1,353 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package container // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/json"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"sync"
+
+	jsoniter "github.com/json-iterator/go"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/entry"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+)
+
+const dockerFormat = "docker"
+const crioFormat = "crio"
+const containerdFormat = "containerd"
+const recombineInternalID = "recombine_container_internal"
+const dockerPattern = "^\\{"
+const crioPattern = "^(?P<time>[^ Z]+) (?P<stream>stdout|stderr) (?P<logtag>[^ ]*) ?(?P<log>.*)$"
+const containerdPattern = "^(?P<time>[^ ^Z]+Z) (?P<stream>stdout|stderr) (?P<logtag>[^ ]*) ?(?P<log>.*)$"
+const logpathPattern = "^.*\\/(?P<namespace>[^_]+)_(?P<pod_name>[^_]+)_(?P<uid>[a-f0-9\\-]+)\\/(?P<container_name>[^\\._]+)\\/(?P<restart_count>\\d+)\\.log$"
+const logPathField = "log.file.path"
+
+var (
+	logFieldsMapping = map[string]string{
+		"stream": "attributes.log.iostream",
+	}
+	k8sMetadataMapping = map[string]string{
+		"container_name": "attributes.k8s.container.name",
+		"namespace":      "attributes.k8s.namespace.name",
+		"pod_name":       "attributes.k8s.pod.name",
+		"restart_count":  "attributes.k8s.container.restart_count",
+		"uid":            "attributes.k8s.pod.uid",
+	}
+)
+
+// Parser is an operator that parses Container logs.
+type Parser struct {
+	helper.ParserOperator
+	recombineParser         operator.Operator
+	format                  string
+	json                    jsoniter.API
+	dockerMatcher           *regexp.Regexp
+	crioMatcher             *regexp.Regexp
+	containerdMatcher       *regexp.Regexp
+	pathMatcher             *regexp.Regexp
+	addMetadataFromFilepath bool
+	crioLogEmitter          *helper.LogEmitter
+	asyncConsumerStarted    bool
+	crioConsumers           sync.WaitGroup
+	parserLock              sync.Mutex
+}
+
+// Process will parse an entry of Container logs
+func (p *Parser) Process(ctx context.Context, entry *entry.Entry) error {
+	format, err := p.detectFormat(entry)
+	if err != nil {
+		return fmt.Errorf("failed to detect a valid container log format", err)
+	}
+
+	if p.format != "" {
+		format = p.format
+	}
```

Comment:
Can we rearrange this so `detectFormat` only runs when the user hasn't specified a format?
```suggestion
	format = p.format
	if format == "" {
		format, err = p.detectFormat(entry)
		if err != nil {
			return fmt.Errorf("failed to detect a valid container log format", err)
		}
	}
```